MaxMind.Db/ArrayBuffer.cs

```diff
@@ -59,6 +79,46 @@ private static byte[] BytesFromStream(Stream stream)
             return bytes;
         }
 
+        private static async Task<byte[]> BytesFromStreamAsync(Stream stream)
+        {
+            if (stream == null)
+            {
+                throw new ArgumentNullException(nameof(stream), "The database stream must not be null.");
+            }
+
+            byte[] bytes;
+
+            // If the stream is within "int" max size, we can read it at once without requiring MemoryStream
+            if (stream.CanSeek && stream.Length <= int.MaxValue)
```

Comment:
These checks need to be removed. We aren't seeking any stream positions and calling stream.Length I believe will go out and buffer entire stream contents to get the length (it isn't cheap). If we are going to do this, we should see if the stream is a MemoryStream first :)

Reply:
This branch was added as part of https://github.com/maxmind/MaxMind-DB-Reader-dotnet/pull/38 - I added the additional check for `bytesRead != stream.Length` on lines 57 and 97 because `Read`/`ReadAsync` aren't guaranteed to fill the buffer with the requested number of bytes (I guess the likelihood of that depends a lot on the type of Stream one passes in).

I'm happy to just use the `MemoryStream` implementation, but I'm curious to hear @maartenba's thoughts on the matter.

Potentially replacing the `ToArray` call with [`TryGetBuffer`](https://msdn.microsoft.com/en-us/library/system.io.memorystream.trygetbuffer.aspx) is a good idea.

Reply:
Yeah, I'm just saying calling length is not a good idea on any stream (minus in memory) as it will cause the whole stream to be consumed to get the length.